api/v1beta2/ratelimitpolicy_types.go

```diff
@@ -0,0 +1,218 @@
+
+	// Rates holds the list of limit rates
+	// +optional
+	Rates []Rate `json:"rates,omitempty"`
+}
+
+// RateLimitPolicySpec defines the desired state of RateLimitPolicy
+type RateLimitPolicySpec struct {
+	// TargetRef identifies an API object to apply policy to.
+	TargetRef gatewayapiv1alpha2.PolicyTargetReference `json:"targetRef"`
+
+	// Limits holds the struct of limits indexed by a unique name
+	// +optional
+	Limits map[string]Limit `json:"limits,omitempty"`
+}
+
+// RateLimitPolicyStatus defines the observed state of RateLimitPolicy
+type RateLimitPolicyStatus struct {
+	// ObservedGeneration reflects the generation of the most recently observed spec.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Represents the observations of a foo's current state.
+	// Known .status.conditions.type are: "Available"
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+}
+
+func (s *RateLimitPolicyStatus) Equals(other *RateLimitPolicyStatus, logger logr.Logger) bool {
+	if s.ObservedGeneration != other.ObservedGeneration {
+		diff := cmp.Diff(s.ObservedGeneration, other.ObservedGeneration)
+		logger.V(1).Info("ObservedGeneration not equal", "difference", diff)
+		return false
+	}
+
+	// Marshalling sorts by condition type
+	currentMarshaledJSON, _ := common.ConditionMarshal(s.Conditions)
+	otherMarshaledJSON, _ := common.ConditionMarshal(other.Conditions)
+	if string(currentMarshaledJSON) != string(otherMarshaledJSON) {
+		diff := cmp.Diff(string(currentMarshaledJSON), string(otherMarshaledJSON))
+		logger.V(1).Info("Conditions not equal", "difference", diff)
+		return false
+	}
+
+	// TODO(eastizle): reflect.DeepEqual does not work well with lists without order
+	if !reflect.DeepEqual(s.GatewaysRateLimits, other.GatewaysRateLimits) {
+		logger.V(1).Info("GatewaysRateLimits not equal")
+		return false
+	}
+	return true
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// RateLimitPolicy is the Schema for the ratelimitpolicies API
+type RateLimitPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   RateLimitPolicySpec   `json:"spec,omitempty"`
+	Status RateLimitPolicyStatus `json:"status,omitempty"`
+}
+
+func (r *RateLimitPolicy) Validate() error {
+	if r.Spec.TargetRef.Group != gatewayapiv1alpha2.Group("gateway.networking.k8s.io") {
+		return fmt.Errorf("invalid targetRef.Group %s. The only supported group is gateway.networking.k8s.io", r.Spec.TargetRef.Group)
+	}
+
+	if r.Spec.TargetRef.Kind != gatewayapiv1alpha2.Kind("HTTPRoute") && r.Spec.TargetRef.Kind != gatewayapiv1alpha2.Kind("Gateway") {
+		return fmt.Errorf("invalid targetRef.Kind %s. The only supported kind types are HTTPRoute and Gateway", r.Spec.TargetRef.Kind)
```

Comment:
Oh, so, we will still only cover the usage of `HTTPRoute` and `Gateway`. I thought for this v2 we were thinking on adding `GatewayClass`.. not sure about the `VirtualService`

Reply:
Currently, it is only HTTPRoute. https://github.com/Kuadrant/kuadrant-operator/issues/156